File: 백준/Silver/1926. 그림/그림.cc
#include <bits/stdc++.h>

using namespace std;

int n, m;
int Max = 0;

int Map[501][501] = { 0, };
int visited[501][501] = { 0, };

int dx[] = { 0, 0, -1, 1 };
int dy[] = { 1, -1, 0, 0 };

int pic_cnt = 0;

typedef struct node
{
	int x;
	int y;
};

queue<node> q;

void bfs()
{
	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j < m; j++)
		{
			if (Map[i][j] == 1 && !visited[i][j])
			{
				pic_cnt++;
				q.push(node{i,j});
				visited[i][j] = 1;

				int cnt = 0;

				while (!q.empty())
				{
					int cx = q.front().x;
					int cy = q.front().y;
					cnt++;
					q.pop();

					for (int k = 0; k < 4; k++)
					{
						int nx = cx + dx[k];
						int ny = cy + dy[k];

						if (nx < 0 || ny < 0 || nx >= n || ny >= m || visited[nx][ny] )
							continue;

						if (Map[nx][ny] == 1)
						{
							q.push(node{nx,ny});
							visited[nx][ny] = 1;
						}
					}
				}
				if (Max < cnt)
					Max = cnt;
			}
		}
	}
}

int main()
{
	cin >> n >> m;

	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j < m; j++)
		{
			cin >> Map[i][j];
		}
	}

	bfs();

	cout << pic_cnt << "\n" << Max;

	return 0;
}

File: 백준/Silver/1926. 그림/그림.java
import java.util.*;
import java.io.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int[][] map;
    static int[][] visited;
    static int[] dx = {0,0,-1,1};
    static int[] dy = {-1,1,0,0};
    static int n, m, cnt, max;

    public static void main(String[] args) throws Exception{
        st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        map = new int[n][m];
        visited = new int[n][m];

        for (int i = 0 ; i < n ; i ++) {
            st = new StringTokenizer(br.readLine());

            for (int j = 0 ; j < m ; j ++) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        for (int i = 0 ; i < n ; i ++) {
            for (int j = 0 ; j < m ; j ++) {
                if (map[i][j] == 1 && visited[i][j] == 0) {
                    cnt ++;
                    bfs(i, j);
                }
            }
        }

        System.out.printf("%d\n%d", cnt, max);
    }

    static void bfs(int x, int y) {
        int max_Cnt = 1;

        ArrayDeque<Node> q = new ArrayDeque<>();
        q.add(new Node(x, y));
        visited[x][y] = 1;

        while(!q.isEmpty()) {
            Node cur = q.poll();

            int cx = cur.x;
            int cy = cur.y;

            for (int i = 0 ; i < 4 ; i ++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if (nx < 0 || ny < 0 || nx >= n || ny >= m || visited[nx][ny] == 1 || map[nx][ny] == 0)
                    continue;

                q.add(new Node(nx, ny));
                visited[nx][ny] = 1;
                max_Cnt ++;
            }
        }

        max = Math.max(max_Cnt, max);

    }

    static class Node{
        int x;
        int y;
        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

}